crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+use super::*;
+use bindings::*;
+use std::sync::*;
+
+/// A type that you can use to declare and implement an event of a specified delegate type.
+///
+/// The implementation is thread-safe and designed to avoid contention between events being
+/// raised and delegates being added or removed.
+pub struct Event<T: Interface + Clone> {
+    swap: Mutex<()>,
+    change: Mutex<()>,
+    delegates: Array<T>,
+}
+
+impl<T: Interface + Clone> Default for Event<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Interface + Clone> Event<T> {
+    /// Creates a new, empty `Event<T>`.
+    pub fn new() -> Self {
+        Self { delegates: Array::new(), swap: Mutex::default(), change: Mutex::default() }
+    }
+    /// Registers a delegate with the event object.
+    pub fn add(&mut self, delegate: &T) -> Result<i64> {
+        let mut _lock_free_drop = Array::new();
+        Ok({
+            let change_lock = self.change.lock().unwrap();
+            let mut new_delegates = Array::with_capacity(self.delegates.len() + 1)?;
+            for delegate in self.delegates.as_slice() {
+                new_delegates.push(delegate.clone());
+            }
+            let delegate = Delegate::new(delegate);
+            let token = delegate.to_token();
+            new_delegates.push(delegate);
+
+            let swap_lock = self.swap.lock().unwrap();
+            _lock_free_drop = self.delegates.swap(new_delegates);
+            token
+        })
+    }
+    /// Revokes a delegate's registration from the event object.
+    pub fn remove(&mut self, token: i64) -> Result<()> {
+        let mut _lock_free_drop = Array::new();
+        {
+            let change_lock = self.change.lock().unwrap();
+            if self.delegates.is_empty() {
+                return Ok(());
+            }
+            let mut capacity = self.delegates.len() - 1;
+            let mut new_delegates = Array::new();
+            let mut removed = false;
+            if capacity == 0 {
+                if self.delegates.as_slice()[0].to_token() == token {
```

Comment:
nit: we might want to just add [indexing](https://doc.rust-lang.org/std/ops/trait.Index.html) directly to the `Array` type